src/sync/sync_permission.hpp

```diff
@@ -0,0 +1,149 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2017 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OS_SYNC_PERMISSION_HPP
+#define REALM_OS_SYNC_PERMISSION_HPP
+
+#include "results.hpp"
+
+namespace realm {
+
+class Permissions;
+class SyncUser;
+class Object;
+
+// Permission object used to represent a user permission.
+// Permission objects can be passed into or returned by various permissions
+// APIs. They are immutable objects.
+struct Permission {
+    // The path of the Realm to which this permission pertains.
+    std::string path;
+
+    // A permission encapsulates a single access level.
+    // Each level includes all the capabilities of the level
+    // above it (for example, 'write' implies 'read').
+    enum class AccessLevel {
+        None,
+        Read,
+        Write,
+        Admin,
+    };
+    AccessLevel access;
+
+    // Condition is a userId or a KeyValue pair
+    // Other conditions may be supported in the future
+    struct Condition {
+        enum class Type {
+            // The permission is applied to a single user based on their user ID
+            UserId,
+            // The permission is based on any user that meets a criterion specified by key/value.
+            KeyValue,
+        };
+        Type type;
+
+        // FIXME: turn this back into a union type
+        std::string user_id = "";
+        std::pair<std::string, std::string> key_value = std::make_pair("", "");
```

Comment:
Why are these initialized like this? The default constructors should be sufficient, no?